## **Pandas Essencial**
**Prof. Dr. Samuel Martins (@hisamuka @xavecoding)** <br/>
xavecoding: https://youtube.com/c/xavecoding <br/>

Neste tutorial, vamos aprender o essencial da biblioteca _Pandas_ para manipulação de dados.<br/><br/>

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

### Pacotes usados neste Notebook

In [1]:
# pacotes usados neste notebook
import pandas as pd

<h1>Manipulação de Dados com Pandas</h1><hr/>

<h2>1. Manipulação Básica de Datasets</h2>
<hr/>

**Dataset**: Gas Prices in Brazil: https://www.kaggle.com/matheusfreitag/gas-prices-in-brazil <br/>

Este dataset contém os **registros dos preços médios semanais dos combustíveis do Brasil entre os anos de 2004 e 2019**. <br/>
Cada *amostra (registro/linha)* consiste em um registro de preço aferido para um dado tipo de combustível em uma dada localidade do Brasil. <br/>
Alguns dos principais *atributos* (colunas) do dataset são: 'ESTADO', 'PRODUTO', 'NÚMERO DE POSTOS PESQUISADOS', 'PREÇO MÉDIO REVENDA'.


\* O arquivo disponibilizado no Kaggle está no formato *tsv*. Embora o _pandas_ consiga abrí-lo normalmente, convertemos tal arquivo para o formato *CSV*, que é um dos formatos mais utilizados, e mudamos seu separado para ';' apenas para mostrar algumas opções da função de carregamento.

### 1.1. Importando o Dataset
Para carregar um dataset no formato csv, basta utilizar a função `read_csv` do pandas. Por padrão, ela considera _','_ como separador.

In [2]:
data = pd.read_csv('./datasets/GasPricesinBrazil_2004-2019.csv')

In [3]:
data

,Unnamed: 0;DATA INICIAL;DATA FINAL;REGIÃO;ESTADO;PRODUTO;NÚMERO DE POSTOS PESQUISADOS;UNIDADE DE MEDIDA;PREÇO MÉDIO REVENDA;DESVIO PADRÃO REVENDA;PREÇO MÍNIMO REVENDA;PREÇO MÁXIMO REVENDA;MARGEM MÉDIA REVENDA;COEF DE VARIAÇÃO REVENDA;PREÇO MÉDIO DISTRIBUIÇÃO;DESVIO PADRÃO DISTRIBUIÇÃO;PREÇO MÍNIMO DISTRIBUIÇÃO;PREÇO MÁXIMO DISTRIBUIÇÃO;COEF DE VARIAÇÃO DISTRIBUIÇÃO;MÊS;ANO
0,0;2004-05-09;2004-05-15;CENTRO OESTE;DISTRITO ...
1,1;2004-05-09;2004-05-15;CENTRO OESTE;GOIAS;ETA...
2,2;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...
3,3;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...
4,4;2004-05-09;2004-05-15;NORDESTE;ALAGOAS;ETANO...
...,...
106818,106818;2019-06-23;2019-06-29;NORDESTE;RIO GRAN...
106819,106819;2019-06-23;2019-06-29;SUL;RIO GRANDE DO...
106820,106820;2019-06-23;2019-06-29;SUL;SANTA CATARIN...
106821,106821;2019-06-23;2019-06-29;SUDESTE;SAO PAULO...


O dataset não foi carregado corretamente pois o separador utilizado seu arquivo CSV era ';' e não a ','. <br/>
Vamos então carregá-lo corretamente:

In [4]:
# Carregando o dataset corretamente ==> neste caso, usa o separador ';'
data = pd.read_csv('./datasets/GasPricesinBrazil_2004-2019.csv', sep=';')

In [5]:
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


### 1.2. Exibindo as primeiras linhas do Dataset
A função `.head()` exibe as 5 primeiras linhas do dataset/tabela/Data Frame.

In [6]:
data.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [7]:
#Exibe as 10 primeiras linhas 
data.head(15)

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
5,5,2004-05-09,2004-05-15,NORDESTE,BAHIA,ETANOL HIDRATADO,408,R$/l,1.383,0.132,...,2.050,0.426,0.095,0.957,0.128,0.5686,1.35,0.134,5,2004
6,6,2004-05-09,2004-05-15,NORDESTE,CEARA,ETANOL HIDRATADO,278,R$/l,1.453,0.218,...,1.950,0.353,0.150,1.1,0.068,0.7332,1.294,0.062,5,2004
7,7,2004-05-09,2004-05-15,NORDESTE,MARANHAO,ETANOL HIDRATADO,105,R$/l,1.631,0.158,...,2.000,0.515,0.097,1.116,0.091,0.9923,1.36183,0.082,5,2004
8,8,2004-05-09,2004-05-15,NORDESTE,PARAIBA,ETANOL HIDRATADO,125,R$/l,1.284,0.130,...,1.699,0.353,0.101,0.931,0.072,0.7577,1.2972,0.077,5,2004
9,9,2004-05-09,2004-05-15,NORDESTE,PERNAMBUCO,ETANOL HIDRATADO,423,R$/l,1.224,0.141,...,1.700,0.277,0.115,0.947,0.091,0.5643,1.3541,0.096,5,2004


### 1.3 Informações do Dataset e Elementos Chave

### 1.3.1 Informações gerais sobre o Dataset

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 21 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Unnamed: 0                     106823 non-null  int64  
 1   DATA INICIAL                   106823 non-null  object 
 2   DATA FINAL                     106823 non-null  object 
 3   REGIÃO                         106823 non-null  object 
 4   ESTADO                         106823 non-null  object 
 5   PRODUTO                        106823 non-null  object 
 6   NÚMERO DE POSTOS PESQUISADOS   106823 non-null  int64  
 7   UNIDADE DE MEDIDA              106823 non-null  object 
 8   PREÇO MÉDIO REVENDA            106823 non-null  float64
 9   DESVIO PADRÃO REVENDA          106823 non-null  float64
 10  PREÇO MÍNIMO REVENDA           106823 non-null  float64
 11  PREÇO MÁXIMO REVENDA           106823 non-null  float64
 12  MARGEM MÉDIA REVENDA          

A primeira coluna da tabela, chamada _'Unnamed: 0'_, parece não significar nada. Na verdade, ela parece ser os **índices** da tabela, que foram salvos como uma _coluna_.<br/>
Veremos jajá como **remover tal coluna**.

Outro ponto é que, aparentemente, nenhum atributo/coluna possui valores nulos (_null_), uma vez que o número de registros do dataframe e os números de valores _non-null_ é de **106823**. <br/>
Mas, veremos que não é bem assim para esse caso.

Além disso, note que alguns atributos/colunas, p. ex., 'PREÇO MÉDIO DISTRIBUIÇÃO', possui tipo de dado `object` ao invés de `float`. Esse `object`, no geral, representa strings. Isso parece estranho e suspeito. Veremos se isso é problemático mais pra frente.

### 1.3.2 Data Frame
Todo dataset carregado (dados estruturados) é um `Data Frame`: 'Tabela' bi-dimensional, de tamanho mutável, com dados potencialmente heterogêneos. <br/>

In [9]:
type(data)

pandas.core.frame.DataFrame

Podemos acessar as **dimensões do Data Frame** (número de linhas x número de colunas) utilizando o atributo `.shape` do Data Frame.

In [10]:
data.shape

(106823, 21)

In [11]:
print(f'O DataFrame possui {data.shape[0]} linhas/observações/registros e {data.shape[1]} colunas/atributos/variáveis.')

O DataFrame possui 106823 linhas/observações/registros e 21 colunas/atributos/variáveis.


#### **Criando um DataFrame**

Podemos criar um DataFrame a partir de um _dicionário_, onde cada **chave** possui uma **lista de elementos de igual tamanho**.<br/>
As **chaves** representam as **colunas** e **cada um dos valores de sua lista** representa o **valor da linha** correspondente para aquela coluna.

In [12]:
personagens_df = pd.DataFrame({
    'nome': ['Luke Skywalker', 'Yoda', 'Palpatine'],
    'idade': [16, 1000, 70],
    'peso': [70.5, 15.2, 60.1],
    'eh jedi': [True, True, False]  # o nome das colunas podem ter espaços
})

In [13]:
personagens_df

,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70.5,True
1,Yoda,1000,15.2,True
2,Palpatine,70,60.1,False


In [14]:
personagens_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   nome     3 non-null      object 
 1   idade    3 non-null      int64  
 2   peso     3 non-null      float64
 3   eh jedi  3 non-null      bool   
dtypes: bool(1), float64(1), int64(1), object(1)
memory usage: 207.0+ bytes


#### **VEJA MAIS**
Criando um Data Frame a partir de um dicionário: https://www.geeksforgeeks.org/how-to-create-dataframe-from-dictionary-in-python-pandas/

#### **Renomeando as colunas de um DataFrame**
**===>** O atributo `DataFrame.columns` retorna uma "lista" com os **nomes de todas as colunas** do data frame.

In [15]:
personagens_df.columns

Index(['nome', 'idade', 'peso', 'eh jedi'], dtype='object')

In [16]:
type(personagens_df.columns)

pandas.core.indexes.base.Index

In [17]:
list(personagens_df.columns)

['nome', 'idade', 'peso', 'eh jedi']

<br/>

**===>** Para **renomear colunas** do data frame, utilize o método `DataFrame.rename`, que retorna uma _cópia_ do data frame com as as colunas renomeadas:

In [18]:
personagens_df

,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70.5,True
1,Yoda,1000,15.2,True
2,Palpatine,70,60.1,False


In [19]:
personagens_df_renomeado = personagens_df.rename(columns={
    'nome': 'Nome Completo', #Renomeia
    'idade': 'Idade'
})

In [20]:
personagens_df_renomeado

,Nome Completo,Idade,peso,eh jedi
0,Luke Skywalker,16,70.5,True
1,Yoda,1000,15.2,True
2,Palpatine,70,60.1,False


In [21]:
personagens_df

,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70.5,True
1,Yoda,1000,15.2,True
2,Palpatine,70,60.1,False


Para renomear o _próprio_ data frame em questão, utilize o parâmetro `inplace=True`:

In [22]:
personagens_df.rename(columns={
    'nome': 'Nome Completo', #Renomeia
    'idade': 'Idade'
}, inplace = True)

In [23]:
personagens_df

,Nome Completo,Idade,peso,eh jedi
0,Luke Skywalker,16,70.5,True
1,Yoda,1000,15.2,True
2,Palpatine,70,60.1,False


<br/>

**===>** Uma outra forma de **renomear todas as listas** de um data frame é passar uma _lista_ com os novos nomes das colunas para o atributo `DataFrame.columns`:

In [24]:
personagens_df.columns

Index(['Nome Completo', 'Idade', 'peso', 'eh jedi'], dtype='object')

In [25]:
personagens_df.columns = ['NOME', 'IDADE', 'PESO', 'EH_JEDI']

In [26]:
personagens_df

,NOME,IDADE,PESO,EH_JEDI
0,Luke Skywalker,16,70.5,True
1,Yoda,1000,15.2,True
2,Palpatine,70,60.1,False


### 1.3.3 Series

Array uni-dimensional com os dados e rótulos de um eixo.

In [27]:
data.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [28]:
#Selecionando uma coluna inteira
data['ESTADO']

0            DISTRITO FEDERAL
1                       GOIAS
2                 MATO GROSSO
3          MATO GROSSO DO SUL
4                     ALAGOAS
                 ...         
106818    RIO GRANDE DO NORTE
106819      RIO GRANDE DO SUL
106820         SANTA CATARINA
106821              SAO PAULO
106822                SERGIPE
Name: ESTADO, Length: 106823, dtype: object

In [29]:
#Selecionando uma coluna inteira 
#Esta forma de acesso só funciona para colunas com nomes sem espaços, acentos etc...
data.ESTADO

0            DISTRITO FEDERAL
1                       GOIAS
2                 MATO GROSSO
3          MATO GROSSO DO SUL
4                     ALAGOAS
                 ...         
106818    RIO GRANDE DO NORTE
106819      RIO GRANDE DO SUL
106820         SANTA CATARINA
106821              SAO PAULO
106822                SERGIPE
Name: ESTADO, Length: 106823, dtype: object

In [30]:
#Selecionando a observação indexada no índice [1] do dataframe
data.iloc[1]

Unnamed: 0                                      1
DATA INICIAL                           2004-05-09
DATA FINAL                             2004-05-15
REGIÃO                               CENTRO OESTE
ESTADO                                      GOIAS
PRODUTO                          ETANOL HIDRATADO
NÚMERO DE POSTOS PESQUISADOS                  387
UNIDADE DE MEDIDA                            R$/l
PREÇO MÉDIO REVENDA                         1.162
DESVIO PADRÃO REVENDA                       0.114
PREÇO MÍNIMO REVENDA                         0.89
PREÇO MÁXIMO REVENDA                        1.449
MARGEM MÉDIA REVENDA                        0.399
COEF DE VARIAÇÃO REVENDA                    0.098
PREÇO MÉDIO DISTRIBUIÇÃO                    0.763
DESVIO PADRÃO DISTRIBUIÇÃO                  0.088
PREÇO MÍNIMO DISTRIBUIÇÃO                  0.5013
PREÇO MÁXIMO DISTRIBUIÇÃO                    1.05
COEF DE VARIAÇÃO DISTRIBUIÇÃO               0.115
MÊS                                             5


In [31]:
type(data.iloc[1])

pandas.core.series.Series

#### **Criando uma Series**

Podemos criar um DataFrame a partir de uma lista de elementos.

In [32]:
pd.Series([5.5, 6.0, 9.5])

0    5.5
1    6.0
2    9.5
dtype: float64

Podemos alterar o **nome dos índices** (veremos melhor jajá) e o **nome da Series** (o que ela representa):

In [33]:
pd.Series([5.5, 6.0, 9.5], index = ['prova 1', 'prova 2','projeto'], name = 'Notas do Luke Skywalker')

prova 1    5.5
prova 2    6.0
projeto    9.5
Name: Notas do Luke Skywalker, dtype: float64

#### **VEJA MAIS**
https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html

### 1.3.4 Atribuindo Dados

In [34]:
data.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


#### 1.3.4.1 Atribuindo constantes

In [35]:
produto_view = data['PRODUTO'] # a series retornada referente à coluna, NÃO É UMA CÓPIA, mas sim uma REFERÊNCIA à coluna do dataframe
produto_view

0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object

In [36]:
produto_copy_bkp = data['PRODUTO'].copy()  # retorna uma cópia da coluna 'PRODUTO'
produto_copy_bkp

0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object

In [37]:
data['PRODUTO'] = 'Combustível'  # atribuindo o valor constante 'Combustível' para linha do dataframe na coluna 'PRODUTO'

In [38]:
data.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,Combustível,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,Combustível,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,Combustível,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,Combustível,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,Combustível,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [39]:
produto_view

0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object

In [40]:
produto_copy_bkp

0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object

#### 1.3.4.2 Atribuindo listas ou series

In [41]:
nrows, ncols = data.shape
nrows, ncols

(106823, 21)

In [42]:
novos_produtos = [f'Produto {i}' for i in range(nrows)]
len(novos_produtos)

106823

In [43]:
# a quantidade de elementos da lista `novos_produtos` é igual ao número de linhas do dataframe
data['PRODUTO'] = novos_produtos

In [44]:
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,Produto 0,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,Produto 1,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,Produto 2,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,Produto 3,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,Produto 4,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,Produto 106818,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,Produto 106819,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,Produto 106820,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,Produto 106821,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


In [45]:
print(produto_view)
print('\n')
print(produto_copy_bkp)


0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object


0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object


In [46]:
# voltando para os produtos originais
data['PRODUTO'] = produto_copy_bkp  # produto_copy_bkp é uma Series

In [47]:
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


#### 1.3.4.3 Criando novas colunas
Para **criar uma nova coluna** em um data frame, basta atribuirmos uma _lista/Series de valores_ a uma **nova 'chave'** do data frame. <br/>

**PS:** A _quantidade de valores_ da lista precisa ser **igual** ao _número de linhas/registros_ do data frame.

In [48]:
# criando uma coluna a partir de um valor constante/default
# todas as linhas terão o mesmo valor para esta nova coluna
data['coluna sem nocao'] = 'DEFAULT'
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,DEFAULT
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004,DEFAULT
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004,DEFAULT
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004,DEFAULT
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004,DEFAULT
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,0.818,0.018,2.756,0,2.756,2.756,0,6,2019,DEFAULT
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019,DEFAULT
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019,DEFAULT
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019,DEFAULT


In [49]:
data['coluna a partir de lista'] = range(data.shape[0])

In [50]:
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,DEFAULT,0
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004,DEFAULT,1
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004,DEFAULT,2
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004,DEFAULT,3
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004,DEFAULT,4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,0.018,2.756,0,2.756,2.756,0,6,2019,DEFAULT,106818
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019,DEFAULT,106819
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,0.065,1.998,0,1.9981,1.9981,0,6,2019,DEFAULT,106820
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019,DEFAULT,106821


In [51]:
# não funciona pq a quantidade de elementos da lista (a serem atribuídos a nova coluna) é diferente
# da quantidade de linhas do dataframe
data['nao funciona'] = [1, 2, 3]

ValueError: Length of values (3) does not match length of index (106823)

<br/>

Outro exemplo:

In [ ]:
data.head()

In [ ]:
data['PREÇO MÉDIO REVENDA (dólares)'] = data['PREÇO MÉDIO REVENDA'] * 6.0

In [ ]:
data

**PS:** Obviamente, a lógica correta em converter o preço dos combustíveis em reais para dólares não é considerar uma taxa de câmbio fixa, uma vez que cada preço foi aferido em um momento diferente.

### 1.3.4 Índices

Todo Data Frame possui **índices**, que não são considerado colunas da tabela. Tais índices são comumente **númericos**, de 0 a num_linhas-1, mas também podem ser **textuais (rótulos/labels)**.

In [ ]:
data

In [ ]:
data.index

Use `list(data.index)` ou `data.index.to_list()` para converter um RangeIndex para uma python list.

#### **Exemplo de Data Frame com Índices Textuais (labels)**

In [ ]:
pesquisa_de_satisfacao = pd.DataFrame({
    'bom': [50, 21, 100],
    'ruim': [131, 2, 30],
    'pessimo': [30, 20, 1]
}, index=['XboxOne', 'Playstation4', 'Switch'])

In [ ]:
pesquisa_de_satisfacao.head()

In [ ]:
pesquisa_de_satisfacao.index

### 1.4 Selecionando uma ou mais amostras (Indexação)

In [ ]:
data.head()

#### **==>  Index-based selection (seleção baseada em Índices)**
Mostrando linhas específicas de um DataFrame:

`iloc`: seleciona elementos do Dataframe, baseado em seu **índice (número)** --> row-first, column-second

**Selecionando uma obsevação/linha:**

In [ ]:
# selecionando a linha 1 ===> observação de índice [1] do dataframe
data.iloc[1]

**Selecionando múltiplas obsevações/linhas:**

In [ ]:
# selecionando as linhas de índice de 0 a 5 (incluso)
data.iloc[:6]

In [ ]:
# selecionando as linhas de índice de 10 a 15 (incluso)
data.iloc[10:16]

In [ ]:
# selecionando as linhas/observações de índice 1, 5, 10, 15
data.iloc[[1, 5, 10, 15]]

In [ ]:
# selecionando as linhas/observações de índices 5, 1, 15, 10
data.iloc[[5, 1, 15, 10]]

In [ ]:
# retornar o valor da linha de índice 1, coluna 4 ('ESTADO')
data.iloc[1, 4]

In [ ]:
# e assim por diante!

#### **==>  Label-based selection (seleção baseadas em Rótulos)**

`loc`: seleciona elementos do Dataframe, baseado em seus **rótulos** --> row-first, column-second

In [ ]:
pesquisa_de_satisfacao

In [ ]:
# retorna a linha de índice 0 (implícito)  ==> usando o iloc
pesquisa_de_satisfacao.iloc[0]

In [ ]:
# retorna o valor da linha de índice 0 (implícito) e coluna de índice 1 (implícito)  ==> usando o iloc
pesquisa_de_satisfacao.iloc[0, 1]

In [ ]:
# retorna a linha cujo o rótulo do índice é 'XboxOne'
pesquisa_de_satisfacao.loc['XboxOne']

In [ ]:
# NÃO FUNCIONA ===> iloc tentando acessar índices com rótulos
pesquisa_de_satisfacao.iloc['XboxOne']

In [ ]:
# NÃO FUNCIONA ===> loc tentando acessar índices com rótulados com números
pesquisa_de_satisfacao.loc[0]

In [ ]:
# retorna o valor da linha 'Playstation4', coluna 'ruim'
pesquisa_de_satisfacao.loc['Playstation4', 'ruim']

In [ ]:
# selecionando as linhas de índices rotulados 'XboxOne', 'Switch'
pesquisa_de_satisfacao.loc[['XboxOne', 'Switch']]

In [ ]:
pesquisa_de_satisfacao.loc[['XboxOne', 'Switch']]

In [ ]:
# retorna todas as linhas e apenas as colunas com rótulos 'bom' e 'pessimo'
pesquisa_de_satisfacao[['bom', 'pessimo']]

In [ ]:
# retorna todas as linhas e apenas as colunas com rótulos 'bom' e 'pessimo'
pesquisa_de_satisfacao.loc[:, ['bom', 'pessimo']]

In [ ]:
#OBSERVAÇÃO

#Índices Númericos, por default, possuem rótulos que correspondem aos seus valores númericos.

In [ ]:
data.head()

In [ ]:
data.loc[1]

### 1.5 Selecionando um ou mais atributos (colunas)

In [ ]:
data.head()

In [ ]:
# retornando a coluna/atributo 'ESTADO'
data['ESTADO']

In [ ]:
data.ESTADO

In [ ]:
data.loc[:, 'ESTADO']

In [ ]:
# a forma de acesso de colunas por .NOME_DA_COLUNA só funciona se
# NOME_DA_COLUNA não possuir caracteres inválidos (espaços, acentos, cedilha, ...)
### não funciona
# data.DATA INICIAL

In [ ]:
# para colunas cujos rótulos possuem caracteres inválidos,
# apenas a seleção via string é válida
data['DATA INICIAL']

In [ ]:
# selecionando múltiplas colunas: 'PRODUTO', 'ESTADO', 'REGIÃO'
data[['PRODUTO', 'ESTADO', 'REGIÃO']]

Como o rótulo da coluna 'DATA INICIAL' **possui espaço**, não é possível acessá-la como método: `data.DATA INICIAL`

### 1.6 Removendo um Atributo (Coluna) do Data Frame 

In [ ]:
data.head()

Como vimos anteriormente, o atributo 'Unnamed: 0' parece ser um **ruído** em nosso dataset. Desta maneira, vamos eliminá-lo,

In [ ]:
# deleta/remove in-place (ou seja, no próprio dataframe) a coluna de rótulo 'Unnamed: 0'
del data['Unnamed: 0']

In [ ]:
data

Vamos agora remover as colunas fictícias criadas nos vídeos para estudo.

In [ ]:
del data['coluna sem nocao']
del data['coluna a partir de lista']
del data['PREÇO MÉDIO REVENDA (dólares)']

In [ ]:
data.head()

### 1.7 Salvando um Data Frame

Para salvarmos um Data Frame para um **arquivo CSV**, basta usarmos o método `.to_csv`. <br/>
Por padrão, esse método **salva os índices da tabela como uma coluna no CSV**.<br/>
Como no geral tais índices são números de 0 a n-1, não há necessidade para isso (veja que removemos anteriormente a coluna 'Unnamed: 0' que foi justamente esse caso).<br/>
Desta forma, utilize o parâmetro: `index=False`.

Por padrão, o método utilizará a ',' como separador das colunas. Caso queira alterá-lo, utilize o parâmetro `sep`.

In [ ]:
data.to_csv('./datasets/GasPricesinBrazil_2004-2019_preprocessado.csv', index=False) #sep=';') 

### 1.8 Seleção Condicional: Filtrando amostras

Durante nossas análise exploratórias, frequentemente filtraremos nossas amostras, a partir de certas **condições**, para fins de análise mais específica. <br/>
Existem algumas maneiras de fazermos tal filtragem. Antes disso, vamos carregar nosso dataset pré-processado que salvamos no item anterior.

In [ ]:
data = pd.read_csv('./datasets/GasPricesinBrazil_2004-2019_preprocessado.csv')

In [ ]:
# Mostra todos os estados cujos os preços dos combustíveis foram aferidos
# Mais tecnicamente, mostra os valores únicos presentes para o atributo/coluna 'ESTADO'.
data['ESTADO'].unique()

#### **Selecionando apenas os preços dos Postos de São Paulo**

##### **==> Alternativa 1: Seleção Condicional (Comparações diretas)**

O código abaixo retorna uma ***Series ('array') de booleans***, com o número de linhas (amostras) do Data Frame, que informa os registros de preços dos postos do _estado de São Paulo_ (True).

In [ ]:
# faz uma comparação elemento a elemento da series, retornando uma Series de booleans
data['ESTADO'] == 'SAO PAULO'

In [ ]:
# salvando essa Series de booleans em uma variável
selecao = data['ESTADO'] == 'SAO PAULO'

In [ ]:
selecao

In [ ]:
type(selecao)

In [ ]:
selecao.shape

In [ ]:
data.shape

Para **filtrarmos** os registros de postos do estado de São Paulo:

In [ ]:
data[selecao]

O resultado é um Data Frame com _apenas_ os registros desejados após a **filtragem**.<br/>
Podemos ainda utilizar o método `loc` para o mesmo fim:

In [ ]:
data.loc[selecao]

##### **==> Alternativa 2: Utilizando o método `query`** 

`query` filtra linhas de um DataFrame baseado em uma **query (pergunta)**.

In [ ]:
data.head()

In [ ]:
data.query('ESTADO == "SAO PAULO"')

<br/>

Uma boa prática é **salvar o Data Frame filtrado em uma nova variável**. Isso simplifica a complexidade do código para futuras análises feita para os postos de São Paulo.

In [ ]:
postos_sp = data.query('ESTADO == "SAO PAULO"')

In [ ]:
postos_sp

In [ ]:
type(postos_sp)

In [ ]:
postos_sp.shape

In [ ]:
postos_sp

Resetando os índices
Note que os índices das linhas/registros após a filtragem continuam sendo os mesmos do DataFrame original.
Em muitas situações, manter esta informação será importante.

Mas, se você quiser resetar os índices do data frame filtrado, fazendo com que os registros começem com índice 0 até num_linhas-1, use o método .reset_index.

In [ ]:
postos_sp.reset_index()

In [ ]:
postos_sp.reset_index(drop=True)

In [ ]:
# ainda não foi alterado, precisamos usar o atributo/parâmetro inplace=True
postos_sp

In [ ]:
postos_sp.reset_index(drop=True, inplace=True)

In [ ]:
postos_sp

In [ ]:
# ALTERNATIVA MUITO COMUM
postos_sp = data.query('ESTADO == "SAO PAULO"').reset_index(drop=True)

In [ ]:
postos_sp

#### **Selecionando registros de postos do Rio de Janeiro com Preços acima de 2 reais**

In [ ]:
 data.head()

In [ ]:
data['ESTADO'] == 'RIO DE JANEIRO'

In [ ]:
data['PREÇO MÉDIO REVENDA'] > 2.0

In [ ]:
selecao = (data['ESTADO'] == 'RIO DE JANEIRO') & (data['PREÇO MÉDIO REVENDA'] > 2.0)
selecao

Note que o resultado da seleção continua sendo uma _Series de booleans_ com o _mesmo número de linhas/observações do DataFrame_, de modo que cada linha possuirá um valor booleano indicando se o posto é do Rio de Janeiro **E** o preço aferido do combustível é maior que 2 reais (True) ou não.

O símbolo **&** significa **AND** na comparação. Essa nomenclatura do python/pandas é diferente das nomenclaturas tradicionais (&&). <br/>
Similarmente:
- **|** representa o **OR** (não é ||)
- **~** representa o **NOT** (não é !)

In [ ]:
data[selecao]

Alternativamente, poderíamos usar o método `query` para fazermos tal seleção. Porém, isso não é possível especificamente para esse caso, pois o rótulo da coluna 'PREÇO MÉDIO REVENDA' possui caracteres inválidos para o método (cedilha, acentos) 

In [ ]:
# Não funciona
# data.query('ESTADO=="RIO DE JANEIRO" and PREÇO MÉDIO REVENDA > 2')

In [ ]:
data.query('ESTADO=="RIO DE JANEIRO" or ESTADO == "SAO PAULO"')

**Aprofundando mais ainda**

A primeira comparação `(data['ESTADO'] == 'RIO DE JANEIRO')` checa, linha a linha (amostra a amostra) do DataFrame, quais são aquelas cujo o estado é RIO DE JANEIRO. Nenhuma averiguação de preços é feita nesse momento. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

A segunda comparação `(data['PREÇO MÉDIO REVENDA'] > 2)` checa, linha a linha (amostra a amostra) do DataFrame, quais são os registros cujo preço do combustível é maior que 2 reais. Note que essa comparação checará os postos de **TODOS os estados**. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

Por fim, as duas "perguntas" são unidas pelo AND (&) que retorna a "pergunta completa" que fizemos.

Alguns podem argumentar que tal abordagem é **ineficiente**, uma vez que, para cada condição ("pergunta"), estamos varrendo todas as linhas do DataFrame. <br/>
O Pandas _tenta otimizar_ isso ao máximo por de trás dos panos. Mas, se fato, de tivermos um dataset **muito grande** (centenas de milhares de linhas), tal abordagem se tornará _lenta_.

Assim, poderíamos fazer filtragem com múltiplos condicionais em partes:

In [53]:
selecao_1 = data['ESTADO'] == 'RIO DE JANEIRO'
postos_rj = data[selecao_1]
postos_rj

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
22,22,2004-05-09,2004-05-15,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,952,R$/l,1.098,0.154,...,0.140,0.786,0.173,0.326,1.263,0.22,5,2004,DEFAULT,22
49,49,2004-05-16,2004-05-22,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,948,R$/l,1.092,0.157,...,0.144,0.808,0.175,0.3257,1.2921,0.217,5,2004,DEFAULT,49
76,76,2004-05-23,2004-05-29,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,973,R$/l,1.103,0.140,...,0.127,0.822,0.166,0.38909,1.1859,0.202,5,2004,DEFAULT,76
103,103,2004-05-30,2004-06-05,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,955,R$/l,1.149,0.156,...,0.136,0.849,0.188,0.3453,1.3667,0.221,5,2004,DEFAULT,103
130,130,2004-06-06,2004-06-12,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,966,R$/l,1.165,0.154,...,0.132,0.911,0.198,0.4,1.3667,0.217,6,2004,DEFAULT,130
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106718,106718,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,ÓLEO DIESEL,212,R$/l,3.569,0.166,...,0.047,3.102,0.119,2.886,3.5162,0.038,6,2019,DEFAULT,106718
106745,106745,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,ÓLEO DIESEL S10,335,R$/l,3.697,0.187,...,0.051,3.197,0.127,2.9772,3.5388,0.04,6,2019,DEFAULT,106745
106772,106772,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,422,R$/l,4.928,0.165,...,0.033,4.39,0.097,4.2,4.618,0.022,6,2019,DEFAULT,106772
106799,106799,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GLP,221,R$/13Kg,65.010,5.812,...,0.089,47.535,3.934,39,58.2,0.083,6,2019,DEFAULT,106799


In [54]:
selecao_2 = postos_rj['PREÇO MÉDIO REVENDA'] > 2
selecao_2

22        False
49        False
76        False
103       False
130       False
          ...  
106718     True
106745     True
106772     True
106799     True
106817     True
Name: PREÇO MÉDIO REVENDA, Length: 4263, dtype: bool

In [55]:
postos_rj_preco_maior_que_2 = postos_rj[selecao_2]
postos_rj_preco_maior_que_2

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
2501,2501,2006-02-26,2006-03-04,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,881,R$/l,2.028,0.148,...,0.073,1.806,0.11,1.31,2.1358,0.061,2,2006,DEFAULT,2501
2528,2528,2006-03-05,2006-03-11,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,889,R$/l,2.125,0.149,...,0.070,1.839,0.117,1.438,2.2801,0.064,3,2006,DEFAULT,2528
2555,2555,2006-03-12,2006-03-18,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,902,R$/l,2.163,0.138,...,0.064,1.895,0.134,1.43,2.3454,0.071,3,2006,DEFAULT,2555
2582,2582,2006-03-19,2006-03-25,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,898,R$/l,2.166,0.146,...,0.067,1.935,0.14,1.438,2.2925,0.072,3,2006,DEFAULT,2582
2609,2609,2006-03-26,2006-04-01,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,890,R$/l,2.172,0.143,...,0.066,1.963,0.133,1.438,2.2513,0.068,3,2006,DEFAULT,2609
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106718,106718,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,ÓLEO DIESEL,212,R$/l,3.569,0.166,...,0.047,3.102,0.119,2.886,3.5162,0.038,6,2019,DEFAULT,106718
106745,106745,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,ÓLEO DIESEL S10,335,R$/l,3.697,0.187,...,0.051,3.197,0.127,2.9772,3.5388,0.04,6,2019,DEFAULT,106745
106772,106772,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,422,R$/l,4.928,0.165,...,0.033,4.39,0.097,4.2,4.618,0.022,6,2019,DEFAULT,106772
106799,106799,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GLP,221,R$/13Kg,65.010,5.812,...,0.089,47.535,3.934,39,58.2,0.083,6,2019,DEFAULT,106799


#### **Selecionando registros de postos de São Paulo ou do Rio de Janeiro com Gasolina Comum acima de 2 reais**

Podemos fazer a solução do "jeito mais lento", percorrendo o DataFrame inteiro _múltiplas vezes_:

In [ ]:
data.head()

In [56]:
selecao_1 = (data['ESTADO'] == 'SAO PAULO') | (data['ESTADO'] == 'RIO DE JANEIRO')
selecao_2 = (data['PRODUTO'] == 'GASOLINA COMUM')
selecao_3 = (data['PREÇO MÉDIO REVENDA'] > 2)

selecao_final = selecao_1 & selecao_2 & selecao_3

In [57]:
selecao_final

0         False
1         False
2         False
3         False
4         False
          ...  
106818    False
106819    False
106820    False
106821    False
106822    False
Length: 106823, dtype: bool

In [63]:
print(data_filtrado['ESTADO'].unique())
print(data_filtrado['PRODUTO'].unique())

['RIO DE JANEIRO' 'SAO PAULO']
['GASOLINA COMUM']


<br/>

Alternativamente:

In [67]:
selecao_1 = (data['ESTADO'] == 'SAO PAULO') | (data['ESTADO'] == 'RIO DE JANEIRO')
postos_sp_rj = data[selecao_1]

# apenas registros de postos dos estados de SP e RJ
postos_sp_rj

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
22,22,2004-05-09,2004-05-15,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,952,R$/l,1.098,0.154,...,0.140,0.786,0.173,0.326,1.263,0.22,5,2004,DEFAULT,22
23,23,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4047,R$/l,0.768,0.085,...,0.111,0.57,0.056,0.4271,1.4123,0.098,5,2004,DEFAULT,23
49,49,2004-05-16,2004-05-22,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,948,R$/l,1.092,0.157,...,0.144,0.808,0.175,0.3257,1.2921,0.217,5,2004,DEFAULT,49
50,50,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,ETANOL HIDRATADO,4066,R$/l,0.766,0.083,...,0.108,0.576,0.05,0.43204,0.9724,0.087,5,2004,DEFAULT,50
76,76,2004-05-23,2004-05-29,SUDESTE,RIO DE JANEIRO,ETANOL HIDRATADO,973,R$/l,1.103,0.140,...,0.127,0.822,0.166,0.38909,1.1859,0.202,5,2004,DEFAULT,76
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106778,106778,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,4.142,0.183,...,0.044,3.713,0.097,3.52,4.0892,0.026,6,2019,DEFAULT,106778
106799,106799,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GLP,221,R$/13Kg,65.010,5.812,...,0.089,47.535,3.934,39,58.2,0.083,6,2019,DEFAULT,106799
106805,106805,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GLP,1017,R$/13Kg,66.893,6.100,...,0.091,49.214,5.071,40.19,72,0.103,6,2019,DEFAULT,106805
106817,106817,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GNV,124,R$/m3,3.111,0.260,...,0.084,2.404,0.174,2.126,2.7909,0.072,6,2019,DEFAULT,106817


In [69]:
selecao_2 = (postos_sp_rj['PRODUTO'] == 'GASOLINA COMUM')
postos_sp_rj_gasolina = postos_sp_rj[selecao_2]

# apenas registros de postos dos estados de SP e RJ cujo produto é GASOLINA COMUM
postos_sp_rj_gasolina

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
12086,12086,2004-05-09,2004-05-15,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1071,R$/l,2.045,0.109,...,0.053,1.791,0.047,1.6363,1.9594,0.026,5,2004,DEFAULT,12086
12087,12087,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,GASOLINA COMUM,4107,R$/l,1.891,0.101,...,0.053,1.656,0.064,1.38,1.944,0.039,5,2004,DEFAULT,12087
12113,12113,2004-05-16,2004-05-22,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1055,R$/l,2.040,0.113,...,0.055,1.788,0.049,1.5943,1.99651,0.027,5,2004,DEFAULT,12113
12114,12114,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,GASOLINA COMUM,4133,R$/l,1.888,0.101,...,0.053,1.655,0.063,1.38,1.9319,0.038,5,2004,DEFAULT,12114
12140,12140,2004-05-23,2004-05-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1091,R$/l,2.042,0.108,...,0.053,1.79,0.046,1.63596,2.2137,0.026,5,2004,DEFAULT,12140
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106479,106479,2019-06-09,2019-06-15,SUDESTE,SAO PAULO,GASOLINA COMUM,1643,R$/l,4.210,0.177,...,0.042,3.785,0.096,3.52,4.1007,0.025,6,2019,DEFAULT,106479
106623,106623,2019-06-16,2019-06-22,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,418,R$/l,4.927,0.160,...,0.032,4.383,0.1,4.1972,4.6239,0.023,6,2019,DEFAULT,106623
106629,106629,2019-06-16,2019-06-22,SUDESTE,SAO PAULO,GASOLINA COMUM,1635,R$/l,4.168,0.177,...,0.042,3.725,0.096,3.502,4.0892,0.026,6,2019,DEFAULT,106629
106772,106772,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,422,R$/l,4.928,0.165,...,0.033,4.39,0.097,4.2,4.618,0.022,6,2019,DEFAULT,106772


In [70]:
postos_sp_rj_gasolina

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
12086,12086,2004-05-09,2004-05-15,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1071,R$/l,2.045,0.109,...,0.053,1.791,0.047,1.6363,1.9594,0.026,5,2004,DEFAULT,12086
12087,12087,2004-05-09,2004-05-15,SUDESTE,SAO PAULO,GASOLINA COMUM,4107,R$/l,1.891,0.101,...,0.053,1.656,0.064,1.38,1.944,0.039,5,2004,DEFAULT,12087
12113,12113,2004-05-16,2004-05-22,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1055,R$/l,2.040,0.113,...,0.055,1.788,0.049,1.5943,1.99651,0.027,5,2004,DEFAULT,12113
12114,12114,2004-05-16,2004-05-22,SUDESTE,SAO PAULO,GASOLINA COMUM,4133,R$/l,1.888,0.101,...,0.053,1.655,0.063,1.38,1.9319,0.038,5,2004,DEFAULT,12114
12140,12140,2004-05-23,2004-05-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1091,R$/l,2.042,0.108,...,0.053,1.79,0.046,1.63596,2.2137,0.026,5,2004,DEFAULT,12140
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106479,106479,2019-06-09,2019-06-15,SUDESTE,SAO PAULO,GASOLINA COMUM,1643,R$/l,4.210,0.177,...,0.042,3.785,0.096,3.52,4.1007,0.025,6,2019,DEFAULT,106479
106623,106623,2019-06-16,2019-06-22,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,418,R$/l,4.927,0.160,...,0.032,4.383,0.1,4.1972,4.6239,0.023,6,2019,DEFAULT,106623
106629,106629,2019-06-16,2019-06-22,SUDESTE,SAO PAULO,GASOLINA COMUM,1635,R$/l,4.168,0.177,...,0.042,3.725,0.096,3.502,4.0892,0.026,6,2019,DEFAULT,106629
106772,106772,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,422,R$/l,4.928,0.165,...,0.033,4.39,0.097,4.2,4.618,0.022,6,2019,DEFAULT,106772


In [71]:
selecao_3 = postos_sp_rj_gasolina['PREÇO MÉDIO REVENDA'] > 2
postos_sp_rj_gasolina_preco_maior_que_2 = postos_sp_rj_gasolina[selecao_3]
postos_sp_rj_gasolina_preco_maior_que_2

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
12086,12086,2004-05-09,2004-05-15,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1071,R$/l,2.045,0.109,...,0.053,1.791,0.047,1.6363,1.9594,0.026,5,2004,DEFAULT,12086
12113,12113,2004-05-16,2004-05-22,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1055,R$/l,2.040,0.113,...,0.055,1.788,0.049,1.5943,1.99651,0.027,5,2004,DEFAULT,12113
12140,12140,2004-05-23,2004-05-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1091,R$/l,2.042,0.108,...,0.053,1.79,0.046,1.63596,2.2137,0.026,5,2004,DEFAULT,12140
12167,12167,2004-05-30,2004-06-05,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1055,R$/l,2.053,0.111,...,0.054,1.802,0.055,1.64,2.21425,0.031,5,2004,DEFAULT,12167
12194,12194,2004-06-06,2004-06-12,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,1091,R$/l,2.063,0.107,...,0.052,1.816,0.054,1.64,2.32425,0.03,6,2004,DEFAULT,12194
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106479,106479,2019-06-09,2019-06-15,SUDESTE,SAO PAULO,GASOLINA COMUM,1643,R$/l,4.210,0.177,...,0.042,3.785,0.096,3.52,4.1007,0.025,6,2019,DEFAULT,106479
106623,106623,2019-06-16,2019-06-22,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,418,R$/l,4.927,0.160,...,0.032,4.383,0.1,4.1972,4.6239,0.023,6,2019,DEFAULT,106623
106629,106629,2019-06-16,2019-06-22,SUDESTE,SAO PAULO,GASOLINA COMUM,1635,R$/l,4.168,0.177,...,0.042,3.725,0.096,3.502,4.0892,0.026,6,2019,DEFAULT,106629
106772,106772,2019-06-23,2019-06-29,SUDESTE,RIO DE JANEIRO,GASOLINA COMUM,422,R$/l,4.928,0.165,...,0.033,4.39,0.097,4.2,4.618,0.022,6,2019,DEFAULT,106772


#### **Selecionando registros dos anos de 2008, 2010 e 2012**

In [72]:
data.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,DEFAULT,0
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004,DEFAULT,1
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004,DEFAULT,2
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004,DEFAULT,3
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004,DEFAULT,4


**ALTERNATIVA 1**

In [73]:
selecao = (data['ANO'] == 2008) | (data['ANO'] == 2010) | (data['ANO'] == 2012)
data[selecao]

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
5098,5098,2008-01-06,2008-01-12,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,80,R$/l,1.806,0.082,...,0.045,1.516,0.061,1.3892,1.6157,0.04,1,2008,DEFAULT,5098
5099,5099,2008-01-06,2008-01-12,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,255,R$/l,1.616,0.100,...,0.062,1.323,0.05,1.2242,1.6173,0.038,1,2008,DEFAULT,5099
5100,5100,2008-01-06,2008-01-12,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,142,R$/l,1.581,0.095,...,0.060,1.303,0.073,1.14,1.45,0.056,1,2008,DEFAULT,5100
5101,5101,2008-01-06,2008-01-12,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,133,R$/l,1.831,0.046,...,0.025,1.449,0.044,1.3247,1.5514,0.03,1,2008,DEFAULT,5101
5102,5102,2008-01-06,2008-01-12,NORDESTE,ALAGOAS,ETANOL HIDRATADO,97,R$/l,1.699,0.098,...,0.058,1.5,0.062,1.39,1.7689,0.041,1,2008,DEFAULT,5102
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
55754,55754,2012-12-30,2013-01-05,NORDESTE,RIO GRANDE DO NORTE,GNV,17,R$/m3,1.891,0.015,...,0.008,1.425,0.098,1.2657,1.4789,0.069,12,2012,DEFAULT,55754
55755,55755,2012-12-30,2013-01-05,SUL,RIO GRANDE DO SUL,GNV,34,R$/m3,1.980,0.040,...,0.020,1.414,0.028,1.3997,1.4608,0.02,12,2012,DEFAULT,55755
55756,55756,2012-12-30,2013-01-05,SUL,SANTA CATARINA,GNV,41,R$/m3,2.003,0.034,...,0.017,-,-,-,-,-,12,2012,DEFAULT,55756
55757,55757,2012-12-30,2013-01-05,SUDESTE,SAO PAULO,GNV,82,R$/m3,1.598,0.199,...,0.125,1.201,0.182,0.8354,1.535,0.152,12,2012,DEFAULT,55757


**ALTERNATIVA 2**

In [76]:
lista_de_anos = [2006,2010,2012]

In [80]:
selecao = data['ANO'].isin(lista_de_anos)
data[selecao]

0         False
1         False
2         False
3         False
4         False
          ...  
106818    False
106819    False
106820    False
106821    False
106822    False
Name: ANO, Length: 106823, dtype: bool

**ALTERNATIVA 3**

In [81]:
lista_de_anos

[2006, 2010, 2012]

In [83]:
data.query('ANO in @lista_de_anos')

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
2263,2263,2006-01-01,2006-01-07,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,199,R$/l,1.817,0.027,...,0.015,1.561,0.094,1.3596,1.9234,0.06,1,2006,DEFAULT,2263
2264,2264,2006-01-01,2006-01-07,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,482,R$/l,1.646,0.090,...,0.055,1.436,0.078,1.2499,1.7199,0.054,1,2006,DEFAULT,2264
2265,2265,2006-01-01,2006-01-07,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,193,R$/l,1.959,0.060,...,0.031,1.556,0.096,1.35,1.759,0.062,1,2006,DEFAULT,2265
2266,2266,2006-01-01,2006-01-07,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,174,R$/l,1.913,0.080,...,0.042,1.627,0.069,1.4905,1.789,0.042,1,2006,DEFAULT,2266
2267,2267,2006-01-01,2006-01-07,NORDESTE,ALAGOAS,ETANOL HIDRATADO,169,R$/l,1.745,0.078,...,0.045,1.526,0.083,1.2722,1.6634,0.054,1,2006,DEFAULT,2267
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
55754,55754,2012-12-30,2013-01-05,NORDESTE,RIO GRANDE DO NORTE,GNV,17,R$/m3,1.891,0.015,...,0.008,1.425,0.098,1.2657,1.4789,0.069,12,2012,DEFAULT,55754
55755,55755,2012-12-30,2013-01-05,SUL,RIO GRANDE DO SUL,GNV,34,R$/m3,1.980,0.040,...,0.020,1.414,0.028,1.3997,1.4608,0.02,12,2012,DEFAULT,55755
55756,55756,2012-12-30,2013-01-05,SUL,SANTA CATARINA,GNV,41,R$/m3,2.003,0.034,...,0.017,-,-,-,-,-,12,2012,DEFAULT,55756
55757,55757,2012-12-30,2013-01-05,SUDESTE,SAO PAULO,GNV,82,R$/m3,1.598,0.199,...,0.125,1.201,0.182,0.8354,1.535,0.152,12,2012,DEFAULT,55757


### **Iterando com DataFrames**

#### For-each `DataFrame.iterrows()` (LENTO ==> apenas indicado para iterar pequenos conjunto de dados)

In [87]:
data.head(10)

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem nocao,coluna a partir de lista
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,DEFAULT,0
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004,DEFAULT,1
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004,DEFAULT,2
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004,DEFAULT,3
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004,DEFAULT,4
5,5,2004-05-09,2004-05-15,NORDESTE,BAHIA,ETANOL HIDRATADO,408,R$/l,1.383,0.132,...,0.095,0.957,0.128,0.5686,1.35,0.134,5,2004,DEFAULT,5
6,6,2004-05-09,2004-05-15,NORDESTE,CEARA,ETANOL HIDRATADO,278,R$/l,1.453,0.218,...,0.150,1.1,0.068,0.7332,1.294,0.062,5,2004,DEFAULT,6
7,7,2004-05-09,2004-05-15,NORDESTE,MARANHAO,ETANOL HIDRATADO,105,R$/l,1.631,0.158,...,0.097,1.116,0.091,0.9923,1.36183,0.082,5,2004,DEFAULT,7
8,8,2004-05-09,2004-05-15,NORDESTE,PARAIBA,ETANOL HIDRATADO,125,R$/l,1.284,0.130,...,0.101,0.931,0.072,0.7577,1.2972,0.077,5,2004,DEFAULT,8
9,9,2004-05-09,2004-05-15,NORDESTE,PERNAMBUCO,ETANOL HIDRATADO,423,R$/l,1.224,0.141,...,0.115,0.947,0.091,0.5643,1.3541,0.096,5,2004,DEFAULT,9


In [86]:
for index, row in data.head(10).iterrows():
    print(f'indice {index} ==> {row["ESTADO"]}')

indice 0 ==> DISTRITO FEDERAL
indice 1 ==> GOIAS
indice 2 ==> MATO GROSSO
indice 3 ==> MATO GROSSO DO SUL
indice 4 ==> ALAGOAS
indice 5 ==> BAHIA
indice 6 ==> CEARA
indice 7 ==> MARANHAO
indice 8 ==> PARAIBA
indice 9 ==> PERNAMBUCO


<h2>2. Preparação dos dados</h2>
<hr/>

### 2.1 Removendo amostras com valores vazios (null / nan) no dataset

De um total de 106823 observações, **não há valores null** / nan para nenhum atributo. Mas, veremos que não é bem assim neste caso específico.<br/><br/>

### 2.2 Conversão de tipos de atributos

O pandas automaticamente reconhece os tipos de dados de cada coluna. <br/>
Porém, existem alguns atributos que estão com seus tipos errados: P. ex., "PREÇO MÉDIO DISTRIBUIÇÃO" deveria ser ```float64``` e não ```object```.<br/>
Nestes casos, muito provavelmente algumas amostras têm um string ao invés de um número para tais atributos. <br/>

Os atributos *"DATA INICIAL"* e *"DATA FINAL"* deveriam ser do tipo `datetime`.

Em outros casos, alguns **atributos categóricos** são ```objects```, mas poderiam ter o tipo ```category```, que é um tipo especial do pandas. <br/>
Este tipo é necessário para se utilizar algumas funções específicas do pandas. <br/>
**Não** converteremos para este tipo por ora.

#### **Datas**
Como os atributos de data do datset já estão em um formato de data aceitável (YYYY-MM-DD), não precisamos forçar nenhuma conversão nesse sentido.

#### **Dados Numéricos**

<br/>

Note que temos vários valores ***null*** agora **após a *conversão de tipos***. Vamos checá-los com mais cuidado nos dados originais e preprocessados.

### 2.3 Limpeza de dados

Várias amostras possuem a _string '-'_ em algumas colunas ao invés de um número de fato. Ou seja, não há aferições destes atributos para estas amostras. <br/>

<br/>

Poderíamos **preencher os valores NaN com um valor padrão**. Para isso, basta usar o método `.fillna`.

<br/>

Por mais que a função `fillna` seja interessante e útil em muitos casos, no problema em questão estamos interessados em analisar precisamente, p. ex., o **'PREÇO MÉDIO DISTRIBUIÇÃO'**.<br/>
A fim de não termos valores _sintéticos_ gerados pelo `fillna`, que possam atrapalhar nossa análise, iremos **remover (drop) todas as amostras que possuem qualquer valor NaN** para quaisquer atributos/colunas. <br/>

Para isso, basta utilizarmos o método `dropna`.

Nosso data frame agora, após essa _limpeza_, ficou mais enxuto, contentdo 103392 registros frente aos 106823 registros originais. <br/>

Essas são apenas algumas das possíveis _técnicas de limpeza de dados_. Outras estratégias, p. ex., **confiam na detecção de outliers**, que não veremos neste curso.

#### **Salvando o Dataset Preprocessado**

<h2>3. Estatísticas Descritivas</h2>
<hr/>

O Pandas fornecem algumas funções/métodos ue computam certas estatísticas descritivas.

`describe`: exibe várias **estatísticas descritivas** para os _atributos_ de um dataframe ou para uma _Series_.

<br/>

Como o resultado do `describe` de um _dataframe_ é outro _dataframe_, podemos filtrar apenas algumas colunas.

**Acessando apenas algumas estatísticas**

<br/>

`mean`, `std`, `min`, etc: cada uma das estatísticas do `describe` podem ser computadas individualmente:

#### Qual é o menor preço mínimo de revenda?

#### Qual é a média e desvio padrão dos preços mínimos de revenda?

#### Quais são os estados considerados?

#### Quantos registros (aferições) cada estado possui?

`.value_counts()`:  Conta a frequência dos valores de uma dada variável (de preferência, _categórica_).

<h2>4. Executando funções para cada item de um DataFrame ou Series</h2>
<hr/>

Uma alternativa ao `for-loop` que vimos anteriormente e que é _lento_, é usarmos _funções próprias do pandas_ que **aplicam/mapeiam uma dada função a todos os elementos de um DataFrame ou Series**, retornando novos elementos "transformados".


<img src='./imagens/apply_map_applymap.png' width=300/>


Fonte: https://towardsdatascience.com/introduction-to-pandas-apply-applymap-and-map-5d3e044e93ff

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4], 
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]}, 
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])

In [ ]:
df

`apply()`: usado para aplicar uma função ao longo de um eixo de um DataFrame ou em valores de uma Series.

<img src='./imagens/pandas_axis.jpg' width=500/>

Fonte: https://www.allthesnippets.com/browse/pandas/df_axis.html

<img src='./imagens/apply_axis_1.png' width=250/>

<img src='./imagens/apply_axis_0.png' width=250/>

##### Usando `lambda` functions

<img src='./imagens/apply_axis_1_mean.png' width=350/>

<br/>

`applymap()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de um DataFrame.

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4], 
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]}, 
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])
df

<br/>

`map()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de uma _Series_.

<h2>5. Agrupamento </h2>
<hr/>

`groupby`: Usado para criar **grupo de elementos** (e.x., baseado nos valores de um atributo). <br/>
**Funções** podem então ser aplicadas para os _elementos de cada grupo_, de modo que os **resultados de cada grupo são combinados**.

<br/>

Também podemos ter agrupamentos por mais de um atributo.

<br/>

`.agg`: **agrega (roda)** uma série de funções para os elementos de um dataframe ou de grupos de um dataframe.

In [ ]:
df = pd.DataFrame([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9],
                   [np.nan, np.nan, np.nan]],
                  columns=['A', 'B', 'C'])
df

<h2>6. Ordenação</h2>
<hr/>

In [ ]:
notas = pd.DataFrame({
    'nome': ['João', 'Maria', 'José', 'Alice'],
    'idade': [20, 21, 19, 20],
    'nota_final': [5.0, 10.0, 6.0, 10.0]
})
notas

`.sort_values()`: ordena valores ao longo de um eixo.

Por padrão, o método retorna uma cópia dos dados ordenados em **ordem crescente (ascendente)**. Podemos alterar isso pelo argumento `ascending`.

<br/>

Podemos ordenar a partir de **mais de uma coluna**:

Ordena os registros, primeiramente, pela coluna 'nota_final' em **ordem descrente**. <br/>
Então, reordena os registros _"empatados"_, ou seja, com a **mesma nota final**, em _ordem alfabética_ (ordem crescente).

<br/>

Note que o dataframe original **não foi alterado** após a ordenação.

Para alterá-lo, use o argumento `inplace=True`:

<h2>7. Exercícios</h2>
<hr/>

Vamos aplicar os conceitos que vimos em alguns exercícios. <br/>
Para isso, utilizaremos o dataset de _preços de combustíveis no Brasil_.

Como há apenas medições de janeiro a junho para o ano de 2019, resolvemos **remover os dados** deste ano da análise.

Note que temos um novo dataframe com 99739 linhas, mas com índices fora desse intervalo. <br/>
Acontece que os registros mantiveram seus índices originais após a query. <br/>

Para resetar os índices de _0 a num_linhas-1_, basta usarmos o método `.reset_index()`. 

Os índices agora foram **resetados**. Porém, os índices antigos se transformaram em _uma nova coluna_ chamada 'index'. <br/>
Para removê-la durante o _reset_, basta passarmos o argumento `drop=True`.

### 7.1 Qual a proporção de postos pesquisados para cada combustível em cada região

### 7.2 Como os preços da Gasolina Comum em São Paulo variaram em 2018?

#### **Estatísticas Descritivas**

### 7.3 Como os preços da Gasolina Comum e do Etanol em São Paulo variaram em 2018?

<h2>8. Assuntos para continuar os estudos</h2>
<hr/>

- join
- concat
- plot
- data cleaning